internal/controllers/cluster/cluster_controller_phases_test.go

```diff
@@ -149,6 +211,201 @@ func TestClusterReconcilePhases(t *testing.T) {
+						},
+					},
+				},
+				expectErr: false,
+			},
+			{
+				name:    "returns no error if the control plane endpoint is not yet set",
+				cluster: clusterNoEndpoint,
+				cpRef: map[string]interface{}{
+					"kind":       "GenericControlPlane",
+					"apiVersion": "controlplane.cluster.x-k8s.io/v1beta1",
+					"metadata": map[string]interface{}{
+						"name":              "test",
+						"namespace":         "test-namespace",
+						"deletionTimestamp": "sometime",
+					},
+					"status": map[string]interface{}{
+						"ready": true,
+					},
+				},
+				expectErr: false,
+			},
+			{
+				name:    "should propagate the control plane endpoint if set",
+				cluster: clusterNoEndpoint,
+				cpRef: map[string]interface{}{
+					"kind":       "GenericControlPlane",
+					"apiVersion": "controlplane.cluster.x-k8s.io/v1beta1",
+					"metadata": map[string]interface{}{
+						"name":              "test",
+						"namespace":         "test-namespace",
+						"deletionTimestamp": "sometime",
+					},
+					"spec": map[string]interface{}{
+						"controlPlaneEndpoint": map[string]interface{}{
+							"host": "example.com",
+							"port": int64(6443),
+						},
+					},
+					"status": map[string]interface{}{
+						"ready": true,
+					},
+				},
+				expectErr: false,
+				check: func(g *GomegaWithT, in *clusterv1.Cluster) {
+					g.Expect(in.Spec.ControlPlaneEndpoint.Host).To(Equal("example.com"))
+					g.Expect(in.Spec.ControlPlaneEndpoint.Port).To(BeEquivalentTo(6443))
+				},
+			},
+			{
+				name:    "should propagate the initialized and ready conditions",
+				cluster: clusterNoEndpoint,
+				cpRef: map[string]interface{}{
+					"kind":       "GenericControlPlane",
+					"apiVersion": "controlplane.cluster.x-k8s.io/v1beta1",
+					"metadata": map[string]interface{}{
+						"name":              "test",
+						"namespace":         "test-namespace",
+						"deletionTimestamp": "sometime",
+					},
+					"spec": map[string]interface{}{},
+					"status": map[string]interface{}{
+						"ready":       true,
+						"initialized": true,
+					},
+				},
+				expectErr: false,
+				check: func(g *GomegaWithT, in *clusterv1.Cluster) {
+					g.Expect(conditions.IsTrue(in, clusterv1.ControlPlaneReadyCondition)).To(BeTrue())
+					g.Expect(conditions.IsTrue(in, clusterv1.ControlPlaneInitializedCondition)).To(BeTrue())
+				},
+			},
+		}
+
+		for _, tt := range tests {
+			t.Run(tt.name, func(t *testing.T) {
+				g := NewWithT(t)
+
+				var c client.Client
+				if tt.cpRef != nil {
+					cpConfig := &unstructured.Unstructured{Object: tt.cpRef}
+					c = fake.NewClientBuilder().
+						WithObjects(builder.GenericControlPlaneCRD.DeepCopy(), tt.cluster, cpConfig).
+						Build()
+				} else {
+					c = fake.NewClientBuilder().
+						WithObjects(builder.GenericControlPlaneCRD.DeepCopy(), tt.cluster).
+						Build()
+				}
+				r := &Reconciler{
+					Client:   c,
+					recorder: record.NewFakeRecorder(32),
+				}
```

Comment:
```suggestion
				r := &Reconciler{
					Client:                    c,
					UnstructuredCachingClient: c,
					recorder:                  record.NewFakeRecorder(32),
				}
```

Looks like we got lucky and we don't hit the panic with our current test cases, but let's make this a bit more future proof :)

Reply:
Ah just looked at the tests now, okay we are hitting the panic :) (would have been surprised otherwise)